jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,297 @@
-                    {
-                        // Restore the demand decremented above.
-                        ++demand;
-                        stalled = true;
-                    }
-                    content.demand(this::process);
-                    return;
-                }
+        // Publisher notes
+        //
+        // 1.6 If a Publisher signals either onError or onComplete on a Subscriber,
+        // that Subscriber’s Subscription MUST be considered cancelled.
+        // 2.4 Subscriber.onComplete() and Subscriber.onError(Throwable t) MUST consider the
+        // Subscription cancelled after having received the signal.
+        //
+        // Publisher failed -> cancel(Throwable)
+        // 1.4 If a Publisher fails it MUST signal an onError.
+        //
+        // Publisher succeeded -> cancel(COMPLETED)
+        // 1.5 If a Publisher terminates successfully (finite stream) it MUST signal an onComplete.
 
-                if (Content.Chunk.isFailure(chunk))
-                {
-                    terminate();
-                    if (!chunk.isLast())
-                        content.fail(chunk.getFailure());
-                    subscriber.onError(chunk.getFailure());
-                    return;
-                }
+        // Subscriber
+        // 2.13 In the case that this rule is violated, any associated Subscription to the Subscriber
+        // MUST be considered as cancelled, and the caller MUST raise this error condition in a
+        // fashion that is adequate for the runtime environment.
+        //
+        // Subscriber.onSubscribe/onNext/onError/onComplete failed -> cancel(new Suppressed(cause))
 
-                subscriber.onNext(chunk);
-                chunk.release();
+        // Subscription notes
+        //
+        // Subscription.cancel -> cancel(new Cancelled())
+        // It's not clearly specified in the specification, but according to:
+        // - the issue: https://github.com/reactive-streams/reactive-streams-jvm/issues/458
+        // - TCK test 'untested_spec108_possiblyCanceledSubscriptionShouldNotReceiveOnErrorOrOnCompleteSignals'
+        // - 1.8 If a Subscription is cancelled its Subscriber MUST eventually stop being signaled.
+        //
+        // Subscription.request with negative argument -> cancel(err)
+        // 3.9 While the Subscription is not cancelled, Subscription.request(long n) MUST signal onError with a
+        // java.lang.IllegalArgumentException if the argument is <= 0.
+    }
 
-                if (chunk.isLast())
-                {
-                    terminate();
-                    // Reactive Stream specification rule 2.9 allows Publishers to call onComplete()
-                    // even without demand, and Subscribers must be prepared to handle this case.
-                    subscriber.onComplete();
-                    return;
-                }
-            }
+    private static class Suppressed extends Throwable
```

Comment:
I'd stick with the common pattern of _not_ extending from `Throwable` (as it is not an `Exception` nor an `Error`), so I think this should be:

```suggestion
    private static class SuppressedException extends Exception
```